odc/geo/cog/_mpu.py

```diff
@@ -240,9 +238,13 @@ def flush(
             bytes_written = self.flush_rhs(write)
 
         if self.left_data:
-            assert len(self.left_data) >= write.min_write_sz
             partId = 1 if leftPartId is None else leftPartId
-            self.parts.insert(0, write(partId, self.left_data))
+            # If finalising, allow flushing leftover even if smaller than min_write_sz.
+            if finalise and len(self.left_data) < write.min_write_sz:
+                self.parts.insert(0, write(partId, self.left_data))
+            else:
+                assert len(self.left_data) >= write.min_write_sz
+                self.parts.insert(0, write(partId, self.left_data))
```

Comment:
NO, can't do that. The assumption is that if you DO keep data on the left, it must be at least as large as a minimal size allowed by the storage backend.

If you are finalizing and you have unflushed left side data, that data needs to be large enough to allow write to the reserved first block, there should not have been any writes to `partId=1` block prior to this.

Reply:
After every merge the following should hold:

- nothing was flushed yet, all data is in memory
- OR
- parts have been written out, and `.left_data` is large enough to be flushed without any padding.

PRIOR to merge assumptions are:

- left side has at least one write credit left so that it can flush `rhs.left_data`

If you are running out of write credits during merge that's already a failure that can not be repaired. I guess we need to make sure to NOT use up last credit in normal operation, if only one write is left then DO NOT flush even if you have enough bytes to make a write, instead just keep appending to in-memory storage and flush it along with `rhs.left_data` when merge happens.